File: src/main/java/edu/metrostate/trackingsystem/infrastructure/utils/PriceDeserializer.java
package edu.metrostate.trackingsystem.infrastructure.utils;

import com.google.gson.*;
import edu.metrostate.trackingsystem.domain.models.Price;

import java.lang.reflect.Type;

public class PriceDeserializer implements JsonDeserializer<Price> {
    @Override
    public Price deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        if (json.isJsonPrimitive() && json.getAsJsonPrimitive().isNumber()) {
            // If the JSON contains only a number, assume default currency as "dollars"
            return new Price(json.getAsDouble(), "dollars");
        } else if (json.isJsonObject()) {
            // If it's a JSON object, parse normally
            JsonObject jsonObject = json.getAsJsonObject();
            double price = jsonObject.has("price") ? jsonObject.get("price").getAsDouble() : 0.0;
            String currency = jsonObject.has("unit") ? jsonObject.get("unit").getAsString() : "dollars";
            return new Price(price, currency);
        }

        throw new JsonParseException("Invalid format for Price field");
    }
}

File: src/main/kotlin/edu/metrostate/dealership/infrastructure/utils/PriceDeserializer.kt
package edu.metrostate.dealership.infrastructure.utils

import com.google.gson.*
import edu.metrostate.dealership.domain.models.Price
import java.lang.reflect.Type

class PriceDeserializer : JsonDeserializer<Price> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Price {
        return when {
            json.isJsonPrimitive && json.asJsonPrimitive.isNumber ->
                Price(json.asDouble, "dollars")

            json.isJsonObject -> {
                val obj = json.asJsonObject
                val price = obj["price"]?.asDouble ?: 0.0
                val currency = obj["currency"]?.asString ?: "dollars"
                Price(price, currency)
            }

            else -> throw JsonParseException("Invalid format for Price field")
        }
    }
}